pkg/preparer/pod_process_reporter/reporter.go

```diff
@@ -0,0 +1,180 @@
+/*
+This package provides a server listening on a unix socket for the exit status
+of pods. The status of each pod is written back to the consul datastore. This
+is useful for allowing external systems to examine the success or failure of a
+pod.
+*/
+package pod_process_reporter
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"io"
+	"net"
+	"os"
+	"time"
+
+	"github.com/square/p2/pkg/kp/statusstore/podstatus"
+	"github.com/square/p2/pkg/launch"
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/types"
+	"github.com/square/p2/pkg/util"
+
+	"github.com/Sirupsen/logrus"
+)
+
+type ReporterConfig struct {
+	// Path to the socket that should receive finish information
+	ProcessResultSocketPath string `yaml:"process_result_socket_path"`
+
+	// Path to the executable that constructs finish information based on process environment
+	EnvironmentExtractorPath string `yaml:"environment_extractor_path"`
+}
+
+type Reporter struct {
+	sockPath                 string
+	environmentExtractorPath string
+
+	logger         logging.Logger
+	podStatusStore podstatus.Store
+}
+
+func New(config ReporterConfig, logger logging.Logger, podStatusStore podstatus.Store) *Reporter {
+	if config.ProcessResultSocketPath == "" {
+		logger.Infoln("process_result_socket_path not configured, process exit status will not be captured")
+		return nil
+	}
+
+	// TODO: do more validation here?
+	if config.EnvironmentExtractorPath == "" {
+		logger.Infoln("environment_extractor_path not configured, process exit status will not be captured")
+		return nil
+	}
+
+	return &Reporter{
+		sockPath:                 config.ProcessResultSocketPath,
+		environmentExtractorPath: config.EnvironmentExtractorPath,
+		logger:         logger,
+		podStatusStore: podStatusStore,
+	}
+}
+
+func (r *Reporter) Run(quitCh <-chan struct{}) error {
+	return r.run(quitCh, 0622)
+}
+
+// This function is broken out so that tests can put more permissive
+// permissions
+func (r *Reporter) run(quitCh <-chan struct{}, socketPerms os.FileMode) error {
+	if err := os.Remove(r.sockPath); err != nil && !os.IsNotExist(err) {
+		return util.Errorf("Could not remove the socket file: %v", err)
+	}
+
+	server, err := net.Listen("unix", r.sockPath)
+	if err != nil {
+		return util.Errorf("could not bind sock: %v", err)
+	}
+
+	r.logger.Infof("Listening for process exit events on %s", r.sockPath)
+	if err := os.Chmod(r.sockPath, socketPerms); err != nil {
+		server.Close()
+		return util.Errorf("could not set FileMode for socket at %s: %v", r.sockPath, err)
+	}
+
+	go func() {
+		<-quitCh
+		server.Close()
+	}()
+
+	for {
+		conn, err := server.Accept()
+		if err != nil {
+			select {
+			case <-quitCh:
+				// We expect an error when server.Close() is called, it's time to shut down
+				r.logger.Infoln("Process event reporter shutting down")
+				return nil
+			default:
+			}
+			r.logger.Errorf("Error during Accept(): %v", err)
```

Comment:
It's here intentionally because we expect an error under normal shutdown (`server.Close()` causes an error in `Accept()`)